### PR TITLE
Migrate buildSrc plugins to build-logic

### PR DESCRIPTION
**Pull request by alexstaeding**
As mentioned in #983, this PR migrates all `buildSrc` plugins to a `build-logic` project.

Everything should be functionally identical to the way it worked previously with the following exceptions:

- The application of the publish plugin is done individually in the api and native projects. Previously, the following if-condition in the velocity publish plugin https://github.com/PaperMC/Velocity/blob/36cc2bde23e7d50545c04b4060a52f3e6b461368/buildSrc/src/main/kotlin/com/velocitypowered/script/VelocityPublishPlugin.kt#L36 would lead to the api and native project to being published. I'm not sure if publishing the native project is intended, but that can be fixed easily.

- I have centralized the repository declaration in `settings.gradle.kts` and enabled `repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)`

### CODE REVIEW

File: build-logic/src/main/kotlin/velocity-publish.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    `maven-publish`
+}
+
+extensions.configure<PublishingExtension> {
+    repositories {
+        maven {
+            credentials(PasswordCredentials::class.java)
+
+            name = "paper"
+            val base = "https://papermc.io/repo/repository/maven"
```
Comment (kashike) [inline build-logic/src/main/kotlin/velocity-publish.gradle.kts]:
```suggestion
            val base = "https://repo.papermc.io/repository/maven"
```

File: settings.gradle.kts [hunk 2]
```diff
@@ -1,3 +1,23 @@
...
+dependencyResolutionManagement {
+    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
+    repositories {
+        mavenCentral()
+        maven("https://s01.oss.sonatype.org/content/repositories/snapshots/") // adventure
+        maven("https://repo.papermc.io/repository/maven-public/")
+    }
+}
+
+
```
Comment (powercasgamer) [inline settings.gradle.kts]:
```suggestion
```

File: build-logic/src/main/kotlin/velocity-publish.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            setUrl(if (version.toString().endsWith("SNAPSHOT")) snapshotsRepoUrl else releasesRepoUrl)
+        }
+    }
+    publications {
+        create<MavenPublication>("maven") {
+            from(components["java"])
+            pom {
+                name.set("Velocity")
+                description.set("The modern, next-generation Minecraft server proxy")
+                url.set("https://www.velocitypowered.com")
```
Comment (powercasgamer) [inline build-logic/src/main/kotlin/velocity-publish.gradle.kts]:
```suggestion
                url.set("https://papermc.io/software/velocity")
```

velocitypowered.com redirects here

**Review by powercasgamer — CHANGES_REQUESTED** [review]

**Comment by alexstaeding** [comment]
Applied suggested fixes for typos and fixed conflicts with base (dev/3.0.0)

**Review by powercasgamer — APPROVED** [review]

**Review by 4drian3d — APPROVED** [review]

**Review by powercasgamer — APPROVED** [review]

**Comment by astei** [comment]
Publishing the native is intentional, it was done for [Krypton](https://github.com/astei/krypton)

**Comment by astei** [comment]
Manually rebased and merged with 24ed83dcadcf55e8831d3f6fa84ab89222f6b222 - thanks for this

**Pull request closed without merging**